### PR TITLE
106 add edit custom field

### PR DESCRIPTION
**Pull request by yaroslavmo**

### CODE REVIEW

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx [hunk 1]
```diff
@@ -33,6 +34,10 @@ class CustomFieldCard extends React.Component<Props> {
       </div>
     );
   }
+
+  private checkDefault = () => {
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx]: rename to `isEditable`

Comment (alexanderko) [reply client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx]: use prototype function

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx [hunk 2]
```diff
@@ -10,7 +11,7 @@ class CustomFieldCard extends React.Component<Props> {
   public render() {
     return (
       <div>
-        <div className={styles.item}>
+        <div className={classNames(styles.item, {[styles.editable]: !this.checkDefault()})}>
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx]: move negation inside function

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.test.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      model: 'Contact',
+      name: 'Address',
+      type: 'string',
+    },
+  ];
+
+  let wrapper: any;
+  const callback = jest.fn();
+  beforeEach(() => {
+    wrapper = shallow(<CustomFieldEditCard title={'Person'} field={customs[1]} saveEdit={callback} cancelEdit={callback} />);
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.test.tsx]: split into lines

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 4]
```diff
@@ -1,12 +1,43 @@
+
+.customFieldsContainer{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 5]
```diff
@@ -1,12 +1,43 @@
...
+  background-color: #f7f7f7;
+  padding-top: 1px;
+  outline: 1px solid #e5e5e5;
+}
+
+.customfieldsWrapper{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 6]
```diff
@@ -1,12 +1,43 @@
...
-  composes: button from "../../../../../../../../styles/common.css";
-  composes: buttonGreen from "../../../../../../../../styles/common.css";
+  composes: button buttonGreen from "../../../../../../../../styles/common.css";
   background-color: #2cbf55;
   font: 600 14px/20px Open Sans, sans-serif;
   margin-bottom: 8px;
   margin-right: 17px;
 }
 
+.button:active{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 7]
```diff
@@ -1,12 +1,43 @@
...
+  background: #25a047;
+  box-shadow: 0 0 0 rgba(38,41,44,0), inset 0 1px 2px rgba(38,41,44,.16);
+}
+
 .buttonWrapper{
   text-align: right;
 }
+
+.addNewField{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 8]
```diff
@@ -1,12 +1,43 @@
...
+  color: #317ae2;
+  font-weight: 600;
+  cursor: pointer;
+  font: 600 14px/20px "Open Sans",sans-serif;
+  margin-left: 16px;
+}
+
+.addNewField:hover{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 9]
```diff
@@ -1,12 +1,43 @@
...
+  text-decoration: underline;
+}
+
+.newFieldContainer{
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: space

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx [hunk 10]
```diff
@@ -2,24 +2,53 @@ import * as React from 'react';
...
+    console.log('Save custom field Action')
+  }
+
+  public cancelEditing = () => {
+    console.log('Cancel editing field Action')
+  }
+
+  public triggerNewField = () => {
+    this.setState((prevState) =>
+      ({ isEdit: !prevState.isEdit })
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx]: use `return` and `{}` brackets

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx [hunk 11]
```diff
@@ -2,24 +2,53 @@ import * as React from 'react';
...
-          return (
-            <CustomFieldCard
-              key={customField._id}
-              customSettings={customField}
-            />);
-        })}
+      <div className={styles.customFieldsContainer}>
+        <div className={styles.customfieldsWrapper}>
+          {this.props.customFields.map(customField => {
+            return (
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx]: don't use return for components

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx [hunk 12]
```diff
@@ -31,6 +60,20 @@ class CustomFields extends React.Component<Props, object> {
...
+      (<CustomFieldEditCard
+        title={this.props.title}
+        saveEdit={this.saveEditing}
+        cancelEdit={this.cancelEditing}
+      />) :
+      (<div className={styles.newFieldContainer}>
+        <span onClick={this.triggerNewField} className={styles.addNewField}>
+          + Add a new field
+    </span>
+      </div>)
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.tsx]: fix indentation

File: client/src/modules/settings/customFieldsActions.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      .then(result => {
+        dispatch({
+          payload: result.data,
+          type: LOAD_DOMAIN,
+        });
+    });
+};
+
+export const editCustomFieldInDomain = 
+  (domainId: string, customField: CustomFieldSetting) =>
```
Comment (alexanderko) [inline client/src/modules/settings/customFieldsActions.ts]: read `domainId` from state

File: client/src/modules/settings/customFieldsActions.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import axios from 'axios';
+import { Dispatch } from 'redux';
+import CustomFieldSetting from '../../models/customFields/CustomFieldSetting';
+import { LOAD_DOMAIN } from './domain/types';
+
+export const addCustomFieldToDomain = 
+  (domainId: string, customField: CustomFieldSetting) =>
```
Comment (alexanderko) [inline client/src/modules/settings/customFieldsActions.ts]: read `domainId` from state

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.css [hunk 15]
```diff
@@ -1,11 +1,66 @@
...
+}
+
+.editButton:active {
+  background: #e5e5e5;
+  box-shadow: inset 0 0 0 1px #cbcccd, 0 0 0 rgba(38, 41, 44, 0), inset 0 1px 2px rgba(38, 41, 44, .16);
+}
+
+.editButton>span {
+  padding: 0 4px;
+  font: 600 12px/16px Open Sans, sans-serif;
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.css]: delete font (redundant)

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.css [hunk 16]
```diff
@@ -33,24 +88,4 @@
...
-  box-shadow: inset 0 0 0 1px #cbcccd, 0 1px 2px rgba(38, 41, 44, .08);
-}
-
-.editButton > span {
-  padding: 3px;
-  font: 600 12px/16px Open Sans, sans-serif;
-  white-space: nowrap;
-  text-overflow: ellipsis;
-}
+}
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.css]: new line

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx [hunk 17]
```diff
@@ -1,38 +1,76 @@
...
 
-class CustomFieldCard extends React.Component<Props> {
+class CustomFieldCard extends React.Component<Props, State> {
+  public state: State = {
+    editMode: false,
+  };
+
+  public editModeHandler = () => {
+    this.setState(prevState => {
+      return {editMode: !prevState.editMode}
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFieldCard/CustomFieldCard.tsx]: spaces in object

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.css [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+  composes: nameInput;
+}
+
+.visibilityOptions {
+  margin-top: 16px;
+  display: grid;
+}
+
+.checkboxLabel {
+  font: 400 14px/20px Open Sans, sans-serif !important;
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.css]: 400 is default

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.css [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+  display: block;
+  position: relative;
+  padding-left: 25px;
+  margin-bottom: 12px;
+  cursor: pointer;
+  font-size: 22px;
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.css]: redundant

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.test.tsx [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        onCancel={callback}
+      />);
+
+    const button = wrapper.find(`#saveButton`);
+    button.simulate('click');
+    expect(callback).toHaveBeenCalled();
+  });
+
+  it('should exec cancelEdit', () => {
+    const button = wrapper.find(`#cancelButton`);
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFieldEditCard/CustomFieldEditCard.test.tsx]: better dont use id only for tests

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 21]
```diff
@@ -1,12 +1,42 @@
...
+.customfieldsWrapper {
+  margin-bottom: 16px;
+}
+
 .button {
-  composes: button from "../../../../../../../../styles/common.css";
-  composes: buttonGreen from "../../../../../../../../styles/common.css";
+  composes: button buttonGreen from "../../../../../../../../styles/common.css";
   background-color: #2cbf55;
   font: 600 14px/20px Open Sans, sans-serif;
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: already exists in button

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 22]
```diff
@@ -1,12 +1,42 @@
...
+  box-shadow: 0 0 0 rgba(38, 41, 44, 0), inset 0 1px 2px rgba(38, 41, 44, .16);
+}
+
+.buttonWrapper {
   text-align: right;
 }
+
+.addNewField {
+  color: #317ae2;
+  font-weight: 600;
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: already exists on font

File: client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css [hunk 23]
```diff
@@ -1,12 +1,42 @@
...
+  margin-left: 16px;
+}
+
+.addNewField:hover {
+  text-decoration: underline;
+}
+
+.newFieldContainer {
+  padding-bottom: 8px;
+}
```
Comment (SamardakOleh) [inline client/src/modules/lead/EditLead/EditLeadSidebar/EditCard/CardFields/EditView/CustomFields/CustomFields.css]: new line

File: client/src/modules/settings/customFieldsActions.ts [hunk 24]
```diff
@@ -3,28 +3,28 @@ import { Dispatch } from 'redux';
...
-      .post(`/api/domain/${getState().auth.domainid}/settings/customFields`, customField)
-      .then(result => {
-        dispatch({
-          payload: result.data,
-          type: LOAD_DOMAIN,
+export const addCustomFieldToDomain =
+  (customField: CustomFieldSetting) =>
+    (dispatch: Dispatch, getState: any) => {
+      axios
+        .post(`/api/domain/${getState().auth.domainid}/settings/customFields`, customField)
```
Comment (alexanderko) [inline client/src/modules/settings/customFieldsActions.ts]: extract `domainId` into variable

File: client/src/modules/settings/customFieldsActions.ts [hunk 25]
```diff
@@ -3,28 +3,28 @@ import { Dispatch } from 'redux';
...
-      .patch(`/api/domain/${getState().auth.domainid}/settings/customFields/${customField._id}`, customField)
-      .then(result => {
-        dispatch({
-          payload: result.data,
-          type: LOAD_DOMAIN,
+export const editCustomFieldInDomain =
+  (customField: CustomFieldSetting) =>
+    (dispatch: Dispatch, getState: any) => {
+      axios
+        .patch(`/api/domain/${getState().auth.domainid}/settings/customFields/${customField._id}`, customField)
```
Comment (alexanderko) [inline client/src/modules/settings/customFieldsActions.ts]: extract `domainId` into variable

**Pull request merged**